### ISSUE TITLE
suggestion: option for `variable-name` rule to allow leading underscore only for unused params

### ISSUE THREAD

**Issue by sethbrenith** [palantir/tslint#3442]
I'm just promoting a comment by @mohsen1 in [this discussion](https://github.com/palantir/tslint/issues/1851) into its own issue for visibility.

Can `variable-name` know about unused params? Most of us don't want to allow leading underscore elsewhere.

```typescript
const _varName = 1; 
//    ~~~~~~~~       <-- this still needs to be a lint error
function foo(_: string, __: number, bar: boolean) {
  console.log('only interested in', bar);
}
```

**Comment by ajafff** [comment]
The latest version of the `naming-convention` rule in my package `tslint-consistent-codestyle` allows you to configure exactly what you want:
https://github.com/ajafff/tslint-consistent-codestyle/blob/master/docs/naming-convention.md#examples
The important part for you is
```js
"naming-convention": [
  true,
  // forbid leading and trailing underscores and enforce camelCase on EVERY name. will be overridden by subtypes if needed
  {"type": "default", "format": "camelCase", "leadingUnderscore": "forbid", "trailingUnderscore": "forbid"},
  ...
  // allow leading underscores for unused parameters, because `tsc --noUnusedParameters` will not flag underscore prefixed parameters
  // all other rules (trailingUnderscore: forbid, format: camelCase) still apply
  {"type": "parameter", "modifiers": "unused", "leadingUnderscore": "allow"}
]
```

**Comment by sethbrenith** [comment]
Excellent, that's just what we were looking for! Thanks!

**Comment by adidahiya** [comment]
Thanks for the link @ajafff. I would like this functionality incorporated into the `variable-name` rule if possible, as an isolated enhancement separate from other features of the `naming-convention` rule.

**Comment by mgol** [comment]
This is now more important as in TypeScript 2.9 `no-unused-variable` is deprecated and if you switch to the compiler `noUnusedLocals` & `noUnusedParameters` options, leading underscores are a hint to the compiler that those params should be allowed to be unused. I'm making such changes to my code base but now the `variable-name` rule complains...

**Comment by sethbrenith** [comment]
@mgol All I see at that link is "allow-leading-underscore". This suggestion was about allowing leading underscores _only_ for unused variables (and still banning them elsewhere).

**Comment by mgol** [comment]
@sethbrenith Ah, sorry, I misinterpreted your request.

**Comment by aryzing** [comment]
@ajafff Since you're most familiar, Would you mind giving this PR a go?

@JoshuaKGoldberg given your participation in #3558, not sure if you're actively looking into this?

**Comment by JoshuaKGoldberg** [comment]
@aryzing no plans on my part to address this issue, but if you want to send a PR that'd be great! 😊

**Comment by pronebird** [comment]
Oh man it would be great to fix this. `@ts-ignore`-ing this is so messed up.

**Comment by JoshuaKGoldberg** [comment]
☠️ *TSLint's time has come!* ☠️ 

TSLint is no longer accepting most feature requests per #4534. See **[typescript-eslint.io](https://typescript-eslint.io)** for the new, shiny way to lint your TypeScript code with **ESLint**. ✨ 

It was a pleasure open sourcing with you all!

**Comment by JoshuaKGoldberg** [comment]
🤖 Beep boop! 👉  **TSLint is deprecated** 👈 _(#4534)_  and you should **switch to [typescript-eslint](https://typescript-eslint.io)**! 🤖 

🔒 This issue is being locked to prevent further unnecessary discussions. Thank you! 👋
